rust/flatc/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! # flatc
+//!
+//! A library for build scripts to compile FlatBuffer schemas.
+//!
+//! This library is intended to be used as a `build-dependencies` entry in
+//! `Cargo.toml`:
+//!
+//! ```toml
+//! [dependencies]
+//! flatbuffers = "0.5"
+//!
+//! [build-dependencies]
+//! flatc = "0.1"
+//! ```
+//!
+//! # Examples
+//!
+//! Use the `Build` structure to compile schema files:
+//!
+//! ```no_run
+//! extern crate flatc;
+//!
+//! fn main() {
+//!     flatc::Build::new()
+//!         .schema("schema/foo.fbs")
+//!         .compile();
+//! }
+//! ```
+//!
+//! Include the generated Rust code:
+//!
+//! ```ignore
+//! mod foo_generated {
+//!     include!(concat!(env!("OUT_DIR"), "/foo_generated.rs"));
+//! }
+//! ```
+
+#![allow(clippy::new_without_default_derive)]
+
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use std::{env, fmt, io};
+
+/// An internal error that occured.
+struct Error {
+    /// The kind of error.
+    kind: ErrorKind,
+    /// A helpful message.
+    message: String,
+}
+
+impl Error {
+    /// Create a new instance of an `Error`.
+    fn new<S: AsRef<str>>(kind: ErrorKind, message: S) -> Self {
+        Self {
+            kind,
+            message: message.as_ref().to_owned(),
+        }
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "An {} error occurred during FlatBuffer compilation: {}",
```

Comment:
How about we add a phrase here to say: "with crate-provided flatc"? Might save users a lot of time debugging flatc versions, especially if they have it installed in `/usr/local/bin` or something, as mac users do.

Reply:
I've extended the error reporting to include the path of the executable compiled by the crate and the commit SHA that it was built from.